sql/src/main/java/io/crate/operation/projectors/ShardingShardRequestConsumer.java

```diff
@@ -129,41 +131,73 @@ public ShardingShardRequestAccumulator(ClusterService clusterService,
     }
 
     @Override
-    public void onItem(Row row) {
+    public CompletableFuture<? extends Iterable<Row>> apply(BatchIterator batchIterator) {
+        CompletableFuture<List<Row>> result = new CompletableFuture<>();
+        consumeIterator(batchIterator, result);
+        return result;
+    }
+
+    private void consumeIterator(BatchIterator batchIterator, CompletableFuture<List<Row>> result) {
+        Row row = RowBridging.toRow(batchIterator.rowData());
+        try {
+            while (batchIterator.moveNext()) {
+                onRow(row);
+                if (idxWithinBatch == batchSize) {
+                    CompletableFuture<BitSet> executeBatchFuture = execute(false);
+                    executeBatchFuture.whenComplete((r, t) -> {
+                        if (t == null) {
+                            consumeIterator(batchIterator, result);
```

Comment:
Couldn't this result in a second loop over the same batchIterator? Concurrent access is not allowed on them.